lib/fir/statements.h

```diff
@@ -0,0 +1,231 @@
+  Variable *variable;
+  TargetListType potentialTargets;
+};
+
+struct UnreachableStmt : public TerminatorStmt_impl {
+  static UnreachableStmt *Create();
+  ~UnreachableStmt() = default;
+
+private:
+  explicit UnreachableStmt() = default;
+};
+
+struct ActionStmt_impl : public Stmt_impl {
+  using ActionTrait = std::true_type;
+  // TODO: DynamicType is a placeholder for now
+  std::optional<evaluate::DynamicType> type_{std::nullopt};
+};
+
+struct AssignmentStmt
+  : public ActionStmt_impl,
+    private SumTypeCopyMixin<std::variant<const parser::AssignmentStmt *,
+        const parser::AssignStmt *>> {
+  friend Statement;
+  ~AssignmentStmt() = default;
+  static AssignmentStmt *Create(const parser::AssignmentStmt *s);
+  static AssignmentStmt *Create(const parser::AssignStmt *s);
+
+private:
+  explicit AssignmentStmt(const parser::AssignmentStmt *s)
+    : SumTypeCopyMixin{s} {}
+  explicit AssignmentStmt(const parser::AssignStmt *s) : SumTypeCopyMixin{s} {}
+};
+
+struct PointerAssignStmt : public ActionStmt_impl {
+  friend Statement;
+  ~PointerAssignStmt() = default;
+  static PointerAssignStmt *Create(const parser::PointerAssignmentStmt *s);
+
+private:
+  explicit PointerAssignStmt(const parser::PointerAssignmentStmt *s)
+    : assign{s} {}
+  const parser::PointerAssignmentStmt *assign;
+};
+
+struct MemoryStmt_impl : public ActionStmt_impl {};
+struct LoadStmt : public MemoryStmt_impl {};
+struct StoreStmt : public MemoryStmt_impl {};
+
+struct AllocateStmt : public MemoryStmt_impl {
+  friend Statement;
+  ~AllocateStmt() = default;
+  static AllocateStmt *Create(const parser::AllocateStmt *s);
+
+private:
+  explicit AllocateStmt(const parser::AllocateStmt *s) : alloc{s} {}
+  const parser::AllocateStmt *alloc;
+};
+
+struct DeallocateStmt : public MemoryStmt_impl {
+  friend Statement;
+  ~DeallocateStmt() = default;
+  static DeallocateStmt *Create(const parser::DeallocateStmt *s);
+
+private:
+  explicit DeallocateStmt(const parser::DeallocateStmt *s) : dealloc{s} {}
+  const parser::DeallocateStmt *dealloc;
+};
+
+struct CastingStmt_impl : public ActionStmt_impl {};
+struct TruncStmt : public CastingStmt_impl {};
+struct ExtendStmt : public CastingStmt_impl {};
+struct FPToIntStmt : public CastingStmt_impl {};
+struct IntToFPStmt : public CastingStmt_impl {};
+struct FPTruncStmt : public CastingStmt_impl {};
+struct FPExtendStmt : public CastingStmt_impl {};
+
+struct CompareStmt_impl : public ActionStmt_impl {};
+
+struct IntCompareStmt : public CompareStmt_impl {};
+
+struct FPCompareStmt : public CompareStmt_impl {};
+
+struct ScopeStmt_impl : public ActionStmt_impl {};
+struct ScopeEnterStmt : public ScopeStmt_impl {};
+struct ScopeExitStmt : public ScopeStmt_impl {};
+struct PHIStmt : public ActionStmt_impl {};
+struct EvalStmt : public ActionStmt_impl {};
+struct CallStmt_impl : public ActionStmt_impl {};
+
+struct CallStmt : public CallStmt_impl {
+  friend Statement;
```

Comment:
Could you please add a comment or two to this file explaining the relationships between the various classes?  I don't know why `Statement` is made a `friend` of all the specific statement classes, or why `CallStmt` inherits from an empty `CallStmt_impl` class.

Reply:
Sorry. I didn't scrub this code at all. I think this is just more left-over scraps from some on-the-fly hacking.